src/crypto-providers/util.ts

```diff
@@ -0,0 +1,265 @@
+): _AddressParameters | null => {
+  const stakePathFile = changeOutputFiles.find(({ path }) => isStakingPath(path))
+  const paymentPathFile = changeOutputFiles.find(({ path }) => !isStakingPath(path))
+  if (!stakePathFile || !paymentPathFile) return null
+  const { pubKey: stakePubKey } = XPubKey(stakePathFile.cborXPubKeyHex)
+  const { pubKey: paymentPubKey } = XPubKey(paymentPathFile.cborXPubKeyHex)
+  const address: Buffer = packBaseAddress(
+    getPubKeyBlake2b224Hash(paymentPubKey),
+    getPubKeyBlake2b224Hash(stakePubKey),
+    network.networkId,
+  )
+  return {
+    address,
+    addressType: getAddressType(address),
+    paymentPath: paymentPathFile.path,
+    stakePath: stakePathFile.path,
+  }
+}
+
+const _packEnterpriseAddress = (
+  changeOutputFile: HwSigningData, network: Network,
+): _AddressParameters => {
+  const { pubKey: paymentPubKey } = XPubKey(changeOutputFile.cborXPubKeyHex)
+  const address: Buffer = packEnterpriseAddress(
+    getPubKeyBlake2b224Hash(paymentPubKey),
+    network.networkId,
+  )
+  return {
+    address,
+    addressType: getAddressType(address),
+    paymentPath: changeOutputFile.path,
+  }
+}
+
+const getChangeAddress = (
+  changeOutputFiles: HwSigningData[],
+  outputAddress: Buffer,
+  network: Network,
+): _AddressParameters | null => {
+  const addressType = getAddressType(outputAddress)
+  try {
+    switch (addressType) {
+      case AddressTypes.BOOTSTRAP:
+        return _packBootStrapAddress(changeOutputFiles[0], network)
+      case AddressTypes.BASE:
+        return _packBaseAddress(changeOutputFiles, network)
+      case AddressTypes.ENTERPRISE:
+        return _packEnterpriseAddress(changeOutputFiles[0], network)
+      default: return null
+    }
+  } catch (e) {
+    return null
+  }
+}
+
+const getAddressAttributes = (address: Address) => {
+  const addressBuffer = addressToBuffer(address)
+  const addressType = getAddressType(addressBuffer)
+  let protocolMagic: ProtocolMagics
+  let networkId: NetworkIds
+
+  if (isValidBootstrapAddress(address)) {
+    protocolMagic = getBootstrapAddressProtocolMagic(addressBuffer)
+    networkId = ProtocolMagics.MAINNET === protocolMagic
+      ? NetworkIds.MAINNET
+      : NetworkIds.TESTNET
+  } else if (isValidShelleyAddress(address)) {
+    networkId = getShelleyAddressNetworkId(addressBuffer)
+    protocolMagic = NetworkIds.MAINNET === networkId
+      ? ProtocolMagics.MAINNET
+      : ProtocolMagics.TESTNET
+  } else throw Error(Errors.InvalidAddressError)
+
+  return { addressType, networkId, protocolMagic }
+}
+
+const ipv4ToString = (ipv4: Buffer | undefined): string | undefined => {
+  if (!ipv4) return undefined
+  return new Uint8Array(ipv4).join('.')
+}
+const ipv6ToString = (ipv6: Buffer | undefined): string | undefined => {
+  if (!ipv6) return undefined
+  const _ipv6LE: Buffer[] = []
+  for (let i = 0; i < 16; i += 4) {
+    _ipv6LE.push(Buffer.from(new Uint32Array([ipv6.readUInt32BE(i)]).buffer))
```

Comment:
Kamil found a more concise way to flip the endianness of the 4-byte chunks: https://www.geeksforgeeks.org/node-js-buffer-swap32-method/

Reply:
Fixed :+1: